mbed_flasher/flashers/FlasherPyOCD.py

```diff
@@ -0,0 +1,202 @@
+        """
+        return PyOCDMap.SUPPORTED_PLATFORMS[platform].get("platform")
+
+    @staticmethod
+    def _get_pack_path():
+        """
+        Return platform specific pack installation dir.
+        Linux: /home/<user>/.local/share/mbed-flasher/pyocd-packs
+        MacOSX: /Users/<user>/Library/Application Support/mbed-flasher/pyocd-packs
+        Windows 7: C:\\Users\\<user>\\AppData\\Local\\mbed-flasher\\mbed-flasher\\pyocd-packs
+        :return: path to packs folder
+        """
+        return path.join(user_data_dir('mbed-flasher'), 'pyocd-packs')
+
+    @staticmethod
+    def pack(platform):
+        """
+        Acquire path to pack file for specific platform
+        :param platform: mbedls board platform
+        :return: Path for pack file, None if not needed
+        raises FlashError if pack file is needed but it doesn't exist
+        """
+        pack_file = PyOCDMap.SUPPORTED_PLATFORMS[platform].get("pack")
+        if pack_file is None:
+            return None
+
+        pack_path = path.join(PyOCDMap._get_pack_path(), pack_file)
+        if path.isfile(pack_path):
+            return pack_path
+
+        raise FlashError(message="Pack file does not exist for the platform: {}".format(platform),
+                         return_code=EXIT_CODE_IMPLEMENTATION_MISSING)
+
+
+class FlasherPyOCD(object):
+    """
+    Flash and erase board using PyOCD.
+    """
+    name = "pyOCD"
+
+    def __init__(self, logger=None):
+        self.logger = logger if logger else logging.getLogger('mbed-flasher')
+
+    @staticmethod
+    def can_flash(target, source):
+        """
+        Function deciding whether the target should be flashed with pyOCD
+        :param target: mbedls given target dictionary
+        :param source: application to be flashed
+        :return: True if target can be flashed with FlasherPyOCD otherwise False
+        """
+        if not source.endswith('.hex'):
+            return False
+
+        return PyOCDMap.is_supported(target["platform_name"])
+
+    @staticmethod
+    def can_erase(target):
+        """
+        Function deciding whether the target should be erased with pyOCD
+        :param target: mbedls given target dictionary
+        :return: True if target can be erased with FlasherPyOCD otherwise False
+        """
+        return PyOCDMap.is_supported(target["platform_name"])
+
+    # pylint: disable=unused-argument
+    def flash(self, source, target, method, no_reset):
+        """Flash target using pyOCD
+        :param source: binary to be flashed
+        :param target: mbedls given target dictionary
+        :param method: method to use when flashing
+        :param no_reset: do not reset flashed board at all
+        :return: 0 if success otherwise raises
+        """
+
+        self.logger.debug('Flashing with pyOCD')
+        session = self._get_session(target, FlashError)
+        try:
+            with session:
```

Comment:
Fyi, _right now_, a **with** statement used on a `Session` will only close the session, not open it. By default, sessions are automatically opened by `ConnectHelper.session_with_chosen_probe()`, although there is an `open_session` bool parameter you can set.

Originally I must have had _some_ reason for doing it this way, but it sure seems dumb now! 😆 So I plan on changing it to make a **with** statement both open and close sessions (or at least figure out why I did it this way, and document that reason).

Reply:
Thanks for the info. Yes, seems more sensible if `with` also opened it.

Reply:
It'll be a little while before I make the change. I just wanted to make clear that there is a corner case with your current code here that could cause the session, and thus the debug probe, to not be closed. But it will be addressed with the change to pyOCD itself.

Otoh, since you have pinned pyOCD to version 0.18.0, maybe you should change the code here. Then when you update pyOCD to a new version that has the `Session` changes, you could revert this code to how it looks right now.

If you do decide to go with the second option, first pass `open_session=False` to `ConnectHelper.session_with_chosen_probe()` in your `_get_session()` method.

Then just call `open()` on the session inside your **with** statement, like this:

```py
session = self._get_session(target, FlashError)
try:
    with session:
        session.open() # *** The new line
        # …
except Blah: #…
```

Of course, you would also make the same change to your `erase()` method.

Reply:
The version is currently pinned to v0.18.0, but we basically can not use this version due to issues found so far.
I'll make the change you suggested as I have understood the `Session` change won't be in the next release at least. Thanks!

Reply:
Changes look good!